compiler/src/iree/compiler/Dialect/Util/IR/UtilOps.td

```diff
@@ -699,15 +702,402 @@ def Util_BufferConstantOp : Util_PureOp<"buffer.constant"> {
+    attr-dict-with-keyword
+  }];
+
+  let extraClassDeclaration = [{
+    Value getOperandSize(unsigned idx) { return getOperandSize(); }
+    Value getResultSize(unsigned idx) { return {}; }
+  }];
+
+  let hasCanonicalizer = 1;
+}
+
+def Util_BufferCopyOp : Util_Op<"buffer.copy", [
+  MemoryEffects<[MemRead, MemWrite]>,
+  Util_SizeAwareOp,
+]> {
+  let summary = [{copies a range of bytes between buffers}];
+  let description = [{
+    Copies a range of bytes as with memcpy (no overlapping).
+  }];
+
+  let arguments = (ins
+    Util_BufferType:$source,
+    Util_Size:$source_size,
+    Util_Offset:$source_offset,
+    Util_BufferType:$target,
+    Util_Size:$target_size,
+    Util_Offset:$target_offset,
+    Util_Size:$length
+  );
+
+  let assemblyFormat = [{
+    $source `[` $source_offset `]` `,`
+    $target `[` $target_offset `]` `,`
+    $length `:`
+    type($source) `` `{` $source_size `}` `->`
+    type($target) `` `{` $target_size `}`
+    attr-dict-with-keyword
+  }];
+
+  let extraClassDeclaration = [{
+    Value getOperandSize(unsigned idx) { return idx == 0 ? getSourceSize() : getTargetSize(); }
+    Value getResultSize(unsigned idx) { return {}; }
+  }];
+
+  let hasCanonicalizer = 1;
+}
+
+def Util_BufferCompareOp : Util_PureOp<"buffer.compare", [
+  MemoryEffects<[MemRead]>,
+  Util_SizeAwareOp,
+]> {
+  let summary = [{compares a range of two buffers}];
+  let description = [{
+    Returns true if the two ranges are bitwise equivalent, somewhat like memcmp.
+  }];
+
+  let arguments = (ins
+    Util_BufferType:$lhs,
+    Util_Size:$lhs_size,
+    Util_Offset:$lhs_offset,
+    Util_BufferType:$rhs,
+    Util_Size:$rhs_size,
+    Util_Offset:$rhs_offset,
+    Util_Size:$length
+  );
+  let results = (outs
+    I1:$result
+  );
+
+  let assemblyFormat = [{
+    $lhs `[` $lhs_offset `]` `,`
+    $rhs `[` $rhs_offset `]` `,`
+    $length `:`
+    type($lhs) `` `{` $lhs_size `}` `,`
+    type($rhs) `` `{` $rhs_size `}`
+    attr-dict-with-keyword
+  }];
+
+  let extraClassDeclaration = [{
+    Value getOperandSize(unsigned idx) { return idx == 0 ? getLhsSize() : getRhsSize(); }
+    Value getResultSize(unsigned idx) { return {}; }
+  }];
+
+  let hasCanonicalizer = 1;
+}
+
+def Util_BufferFillOp : Util_Op<"buffer.fill", [
+  MemoryEffects<[MemWrite]>,
+  Util_SizeAwareOp,
+]> {
+  let summary = [{fills a range of bytes with a value}];
+  let description = [{
+    Fills the contents of the buffer in the given byte range with a pattern.
+    The offset and length must match the natural alignment of the pattern type.
+  }];
+
+  let arguments = (ins
+    Util_FillPattern:$pattern,
+    Util_BufferType:$target,
+    Util_Size:$target_size,
+    Util_Offset:$target_offset,
+    Util_Size:$length
+  );
+
+  let assemblyFormat = [{
+    $pattern `,`
+    $target `[` $target_offset `for` $length `]` `:`
+    type($pattern) `->`
+    type($target) `` `{` $target_size `}`
+    attr-dict-with-keyword
+  }];
```

Comment:
"[offset] for [length]" here is a little strange IMO. I could get used to it, but it might be more clear to spell out "offset" and/or "length" in the assembly format, instead of abbreviating this much. Do any other IREE / upstream ops use "for" in this way in assembly formats?

```mlir
// current
util.buffer.fill %arg2, %arg0[%c100 for %c200] : i32 -> !util.buffer{%arg1}

// suggestion
util.buffer.fill %arg2, %arg0[offset: %c100, length: %c200] : i32 -> !util.buffer{%arg1}
```

Reply:
yeah flow, stream, and util dialect use this - if we change it we'd want to change all of them